scripts/crowbar-testbuild.py

```diff
@@ -100,59 +109,109 @@ def add_pr_to_checkout(repo, pr_id, head_sha1, pr_branch, spec):
+             '--noservice',
+             repository, 'x86_64', spec,
+             _out=tee)
+    finally:
+        out.close()
+
+    sh.cp('-p',
+          sh.glob(os.path.join(buildroot,
+                               '.build.packages/RPMS/*/*.rpm')),
+          webroot)
+
 
-    olddir = os.getcwd()
+def trigger_testbuild(args):
     workdir = tempfile.mkdtemp()
     build_failed = False
+    pkg = args.repo if args.repo == "crowbar" else "crowbar-" + args.repo
+    spec = pkg + '.spec'
+    ptfdir = ':'.join([args.repo, args.pr_id, args.sha1, args.branch])
+    ptfdir, n = re.subn('/', '_', ptfdir)
+    webroot = prep_webroot(ptfdir)
+
+    ptf_url = htdocs_url + ptfdir
+    pr_set_status = \
+        functools.partial(ghs_set_status,
+                          args.org, args.repo, args.sha1, ptf_url)
+
+    pr_set_status('pending', 'building PTF package')
+
     try:
-        ptfdir = repo + ':' + github_opts
-        webroot = os.path.join(htdocs_dir, ptfdir)
-        pkg = repo if repo == "crowbar" else "crowbar-" + repo
-        spec = pkg + '.spec'
-
-        sh.rm('-rf', webroot)
-        sh.mkdir('-p', webroot)
-        try:
-            os.chdir(workdir)
-            buildroot = os.path.join(os.getcwd(), 'BUILD')
-            iosc('co', IBS_MAPPING[pr_branch], pkg, '-c')
-            os.chdir(pkg)
-            add_pr_to_checkout(repo, pr_id, head_sha1, pr_branch, spec)
-            repository = 'SLE_12' if pr_branch == 'master' else 'SLE_11_SP3'
-            iosc('build', '--root', buildroot, '--noverify', '--noservice',
-                 repository, 'x86_64', spec, _out=sys.stdout)
-        except:
-            build_failed = True
-            print("Build failed: " + str(sys.exc_info()[0]))
-            raise
-        else:
-            sh.cp('-p', sh.glob(
-                os.path.join(buildroot, '.build.packages/RPMS/*/*.rpm')),
-                webroot)
-        finally:
-            os.chdir(olddir)
-            sh.cp('-p', os.path.join(buildroot, '.build.log'),
-                  os.path.join(webroot, 'build.log'))
+        prep_osc_dir(workdir,
+                     args.org, args.repo, args.pr_id, args.sha1, args.branch,
+                     pkg, spec)
+        build_package(spec, webroot, args.branch)
+    except:
+        build_failed = True
+        exc_type, exc_val, exc_tb = sys.exc_info()
+        msg = "Build failed: %s" % exc_val
+        print(msg)
+        with open(os.path.join(webroot, 'build.err'), 'w') as err:
+            err.write(msg)
     finally:
         sh.sudo.rm('-rf', workdir)
 
-    if not build_failed:
-        jenkins_job_trigger(
-            repo, github_opts, CLOUDSRC[pr_branch], ptfdir)
+    if build_failed:
+        pr_set_status('failure', 'PTF package build failed')
+    else:
+        result = jenkins_job_trigger(
+            args.org, args.repo, args.pr_id, args.sha1,
+            CLOUDSRC[args.branch], ptf_url)
+        print(result)
+        pr_set_status('pending')
+
+
+def main():
```

Comment:
> seriously? for a main body that just calls trigger_testbuild?

It doesn't "just" call `trigger_testbuild()`, it also does all the CLI handling.  That's an important line to draw between the UI and non-UI code, which will also make it easier in the future if we ever want to use this as a module as well as a script.

> you could have just renamed trigger_testbuild to main()..

Nope because then there would have been a bunch of code outside `main()`.

This should not be a controversial change - it's the industry standard approach, e.g.

- http://stackoverflow.com/a/20158605/179332
- https://google-styleguide.googlecode.com/svn/trunk/pyguide.html#Main
- https://www.chromium.org/chromium-os/python-style-guidelines#TOC-Main-Modules

Furthermore the vast majority of OpenStack executables adopt the same convention:

    rpm -qa | grep openstack | xargs rpm -ql | grep /usr/bin | xargs grep -A1 __main__